QEfficient/lora/auto.py

```diff
@@ -0,0 +1,366 @@
+
+                # stack weight tensors
+                stacked_lora_a = (
+                    torch.stack(a_tensor_list, dim=0).unsqueeze(1).transpose(2, 3)
+                )  # <num_loras, 1, in_feature, r>
+                stacked_lora_b = (
+                    torch.stack(b_tensor_list, dim=0).unsqueeze(1).transpose(2, 3)
+                )  # <num_loras, 1, r, out_feature>
+                stacked_lora_s = (
+                    torch.stack(s_tensor_list, dim=0).unsqueeze(1).unsqueeze(2).unsqueeze(3)
+                )  # <num_loras, 1, 1, 1>
+
+                # stored weight to corresponding ops
+                if target_module == "q_proj":
+                    module = self.model.model.layers[i].self_attn.q_proj
+                elif target_module == "k_proj":
+                    module = self.model.model.layers[i].self_attn.k_proj
+                elif target_module == "v_proj":
+                    module = self.model.model.layers[i].self_attn.v_proj
+                elif target_module == "o_proj":
+                    module = self.model.model.layers[i].self_attn.o_proj
+                else:
+                    raise NotImplementedError("Target module not supported!!")
+
+                module.lora_a_weights.copy_(stacked_lora_a)
+                module.lora_b_weights.copy_(stacked_lora_b)
+                module.lora_scalings.copy_(stacked_lora_s)
+
+    def _init_adapter_model(self):
+        "Initialize the fixed lora model with multiple adapter weigths standby"
+
+        # assume all adapters have same target_modules and ranks
+        if self.max_num_adapters != len(self.active_adapter_to_id):
+            raise ValueError("Inconsistent max_num_adapters and active adapters")
+
+        # set lora rank
+        self.lora_rank = list(self.adapter_configs.values())[0].r
+
+        # do the module replacement
+        _, transformed = LoraModelInputsTransform.apply(self.model)
+
+        self.target_modules_for_all_adapters = list(self.adapter_configs.values())[0].target_modules
+        _, transformed = TargetModulesTransform.apply(
+            self.model, self.target_modules_for_all_adapters, self.lora_rank, self.max_num_adapters
+        )
+
+        # load_weight to model
+        self._load_adapter_weights_to_model()
+
+    def export(self, export_dir: Optional[str] = None) -> str:
+        """
+        Exports the model to ``ONNX`` format using ``torch.onnx.export``.
+        We currently don't support exporting non-transformed models. Please refer to the ``convert_to_cloud_bertstyle`` function in the **Low-Level API** for a legacy function that supports this."
+
+        ``Optional`` Args:
+            does not any arguments.
+
+        Returns:
+            :str: Path of the generated ``ONNX`` graph.
+        """
+
+        # initialize the adapter model
+        assert self.max_num_adapters, "Please use load_adapter() to add at least one adapter; otherwise, refer to QEFFAutoModelForCausalLM for base model usage"
```

Comment:
please raise a ValueError instead of assert.

Reply:
Updated.